lib/handler/RetryHandler.js

```diff
@@ -0,0 +1,324 @@
+const assert = require('node:assert')
+
+const { kRetryHandlerDefaultRetry } = require('../core/symbols')
+const { RequestRetryError } = require('../core/errors')
+const {
+  isDisturbed,
+  parseHeaders,
+  parseRangeHeader,
+  safeHTTPMethods
+} = require('../core/util')
+
+class RetryHandler {
+  constructor (opts, handlers, retryOpts) {
+    const {
+      // Retry scoped
+      retry: retryFn,
+      max,
+      maxTimeout,
+      minTimeout,
+      timeoutFactor,
+      // Response scoped
+      methods,
+      idempotent,
+      codes,
+      status: statusCodes
+    } = retryOpts ?? {}
+
+    this.dispatch = handlers.dispatch
+    this.handler = handlers.handler
+    this.opts = opts
+    this.abort = null
+    this.aborted = false
+    this.retryOpts = {
+      retry: retryFn ?? RetryHandler[kRetryHandlerDefaultRetry],
+      maxTimeout: maxTimeout ?? 30 * 1000, // 30s,
+      timeout: minTimeout ?? 500, // .5s
+      timeoutFactor: timeoutFactor ?? 2,
+      max: max ?? 8,
+      // What errors we should retry
+      methods: methods ?? ['GET', 'HEAD', 'OPTIONS', 'PUT', 'DELETE', 'TRACE'],
+      // States weather or not retry on idempotent methods
+      idempotent: idempotent ?? false,
+      // Indicates which errors to retry
+      status: statusCodes ?? [500, 502, 503, 504],
```

Comment:
why no 429?

Reply:
I haven't implemented that one yet; as its nature is coming from a rate-limit logic, we need to handle its `Retry-After` header; I'm planning to add settings specifically to it and its own default logic